### PR TITLE
Restart operation on child-devices

### PR DESCRIPTION
**Pull request by didier-wenzek**
Updating `tedge-agent` and `c8y-mapper` to support restart operation on child devices

## Proposed changes

- [x] use `mqtt.topic_root` and `mqtt.device_topic_id`
- [x] use ` te/device/main///cmd/restart/+` instead of `tedge/commands/+/control/restart`
- [x] move the command id from the payload to the topic
- [x] publish restart commands as retained messages
- [x] The c8y-mapper has to register child device restart capability
- [x] Add system tests: restarting a child device from c8y

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/2018

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by reubenmiller** [comment]
@didier-wenzek is the PR in a state where we can add an integration test for the child device as well? Or is there still other work required for the tedge-agent to function on a child device (in the scope of the restart command only)

File: crates/core/tedge_agent/src/file_transfer_server/http_rest.rs
Comment (albinsuresh) [inline crates/core/tedge_agent/src/file_transfer_server/http_rest.rs:351]: Any specific issue that triggered this change?

File: crates/core/tedge_api/src/messages.rs
Comment (albinsuresh) [inline crates/core/tedge_api/src/messages.rs:35]: Looked bold, but makes sense.

File: crates/core/tedge_api/src/messages.rs
Comment (albinsuresh) [inline crates/core/tedge_api/src/messages.rs:494]: ~Wouldn't it be better to pass that `cmd_id` as an argument, so that the caller has better control on whether to create unique IDs or pass something know? Especially for unit tests.~

Saw the `with_id` function below that solves this problem. This is better.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:998]: I just realised that you can have multiple plain impl blocks for the same struct, which are not for any trait implementations and this module already had 3 different impl blocks. Looks fine, but I just don't understand the benefits though. I'm assuming it is for better code organization to isolate different responsibilities into different impl blocks instead of one big fat impl block.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:1013]:
```suggestion
                        error!("Unsupported `restart` operation for a service: {target}");
```

**Review by albinsuresh — APPROVED** [review]
LGTM. I was surprised to see that no integration tests were updated. But then found out that none exists atm. I'm hoping that it is feasible now, since we have containers in place of real devices.

File: crates/core/tedge_agent/src/file_transfer_server/http_rest.rs
Comment (reubenmiller) [reply crates/core/tedge_agent/src/file_transfer_server/http_rest.rs:351]: Looking at the individual commits is helpful here ;)

Comment (didier-wenzek) [reply crates/core/tedge_agent/src/file_transfer_server/http_rest.rs:351]: See https://github.com/thin-edge/thin-edge.io/pull/2245/commits/4874c95f7e159d99510a6735c2fdb6a736399261

File: crates/core/tedge_api/src/messages.rs
Comment (didier-wenzek) [reply crates/core/tedge_api/src/messages.rs:35]: I decided to move that way when I saw the impact for nothing on so many places: https://github.com/thin-edge/thin-edge.io/pull/2245/commits/162a83ba520ccd4862874b1d7eed7b0d89cc4a23.

**Comment by reubenmiller** [comment]
This might be a bit premature, but I did a quick manual test to see if I could trigger a restart operation from Cumulocity on the child device, but
* c8y_Restart operation is not registered on the c8y device managed object
* the tedge-agent running on the child device does not respond to the operation, the operation stays in PENDING status in Cumulocity

So if this is not the expected behaviour (due to that part not being implemented yet), then I would suggest we wait to add some new integration tests for this new functionality before we merge it.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs:998]: Indeed, something that needs to be fixed. But in a PR doing only that.

Why are we here? Previously, there was an implementations of a trait used *only* here. There were also functions which have been promoted to methods. When I introduced [these changes](https://github.com/thin-edge/thin-edge.io/pull/2108), I decided to go that way in order to help the code reviewers by minimizing code shuffling on an already huge refactoring.

Now, this can be cleaned up.

**Comment by didier-wenzek** [comment]
> So if this is not the expected behaviour (due to that part not being implemented yet), then I would suggest we wait to add some new integration tests for this new functionality before we merge it.

Sure, we must add system tests with the 3 systems (c8y, c8y-mapper and tedge-agent). The integration tests are only checking partial exchanges between c8y-mapper and tedge-agent.

**Comment by reubenmiller** [comment]
> > So if this is not the expected behaviour (due to that part not being implemented yet), then I would suggest we wait to add some new integration tests for this new functionality before we merge it.
> 
> Sure, we must add system tests with the 3 systems (c8y, c8y-mapper and tedge-agent). The integration tests are only checking partial exchanges between c8y-mapper and tedge-agent.

Then I'll add the system tests on Monday, though we definitely have issues here (as previously posted).

**Comment by didier-wenzek** [comment]
> @didier-wenzek is the PR in a state where we can add an integration test for the child device as well? Or is there still other work required for the tedge-agent to function on a child device (in the scope of the restart command only)

The PR should be in a state where a restart of the child-device should work.

> I was also surprised to see that no integration tests were updated. But then found out that none exists atm. I'm hoping that it is feasible now, since we have containers in place of real devices.

There is at least this one for the main device:

https://github.com/thin-edge/thin-edge.io/blob/main/tests/RobotFramework/tests/cumulocity/restart/restart_device.robot

**Comment by albinsuresh** [comment]
> > I was also surprised to see that no integration tests were updated. But then found out that none exists atm. I'm hoping that it is feasible now, since we have containers in place of real devices.
> 
> There is at least this one for the main device:
> 
> https://github.com/thin-edge/thin-edge.io/blob/main/tests/RobotFramework/tests/cumulocity/restart/restart_device.robot

Aah okay, using the Cumulocity APIs instead of triggering commands via `te/.../cmd/restart` topics. I was searching for a direct test using the `tedge/` or `te/` topics and that's why I didn't get any results.

**Comment by didier-wenzek** [comment]
This PR has been rebased on top of https://github.com/thin-edge/thin-edge.io/pull/2265

**Review by Bravo555 — APPROVED** [review]

**Comment by didier-wenzek** [comment]
I confirm something is going wrong :-(

Restart operations are pushed and executed but then the agent start looping while the mapper say to nothing to c8y.

```
```

**Comment by didier-wenzek** [comment]
To be fixed:
- [ ] the agent must ignore commands which are not in the init state
- [ ] the mapper should use better cmd_id e.g. `c8y-{timestamp}`
- [ ] the mapper correctly send `[c8y/s/us] 502,c8y_Restart,"Restart Failed"` but this is ignored by the cloud. Why?
- [ ] the mapper must clear successful and failed requests
- [ ] the reason for a restart failure is not given
- [ ] on restart, the mapper should not create a new restart request (when pending in the cloud and locally) 
- [ ] it would be helpful to add to each message the name of the service which changed the state

**Comment by didier-wenzek** [comment]
> The reason behind the status updates and failure reason not working is because they're being published on the incorrect c8y topic. For child devices it should be c8y/s/us/{child_id}

I started by testing the main device.

**Comment by reubenmiller** [comment]
@didier-wenzek I've added a system test to the PR

File: tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot
Comment (didier-wenzek) [inline tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot:14]: @reubenmiller Where is this check defined?

File: tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot
Comment (didier-wenzek) [inline tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot:61]: Support of `c8y_LogfileRequest` is explicitly added here. Why? I would expect to have this added when `tedge_log_plugin` starts.

Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot:61]: Sorry that is my mistake (copy paste error). Maybe this is the source of the error. When I wrote the system tests for the tedge-log-plugin the supported operations mechanism was not implemented, so I compensated for it...I'll double check if this is the case now or not and update the tests accordingly.

File: tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot
Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot:14]: The assertion checks if the `c8y_Restart` operation type has been declared in Cumulocity IoT REST API.

The capabilities are registered fine on the tedge broker, however they are definitely not translated to cloud messages. I believe this functionality is missing as the c8y_LogFileRequest operation is also not sent to the cloud.

File: tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot
Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot:61]: I've repeated the checks, and tedge does not forward the locally registered capabilities to Cumulocity...so this functionality probably hasn't been implemented yet.

Comment (didier-wenzek) [reply tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot]: I'm missing something. I tested manual registration is working as I was expecting.

1. On start, `tedge-agent` publishes its capabilities `[te/device/child777-002///cmd/restart] {}`
2. On reception of this message, `c8y-mapper` publishes an auto-registration messages : `[te/device/child777-002//] { "@type":"child-device", "@id":"child777-002"}`.
3. On reception of this message, `c8y-mapper` also registers the child device and its capabilities:
     1. `[c8y/s/us] 101,child777-002,child777-002,thin-edge.io-child`
     2. `[c8y/s/us/child777-002] 114,c8y_Restart`

Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot]: Strange. I'll have another look, and try and strip back the test setup to a bare minimum, though I believe the test was already fairly minimal...

Comment (didier-wenzek) [reply tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot]: > though I believe the test was already fairly minimal...

The test looks good. However, it's not obvious for me to see which checks are running on the main device and which on the child device.

Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot]: Ok, I've fixed it...the child device and capability are being registered correctly without doing any manual file creation (which what the test was previously doing).

So the only thing which is broken is the updating of the c8y_Restart operation status when operating on a child device.

Comment (didier-wenzek) [reply tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot]: > Ok, I've fixed it...the child device and capability are being registered correctly without doing any manual file creation (which what the test was previously doing).

A file is created by the c8y mapper on the main device.

> So the only thing which is broken is the updating of the c8y_Restart operation status when operating on a child device.

Thanks. I will have a look.

File: crates/core/tedge_agent/src/restart_manager/actor.rs
Comment (reubenmiller) [inline crates/core/tedge_agent/src/restart_manager/actor.rs:202]: I don't think we should treat this as an error here as the script/process can be stopped via a signal as a result of the reboot handling (e.g. systemd killing the services quickly after `shutdown -r now` or `reboot` is executed in the command).

Instead we can just treat it like it was successful, and let the failed restart logic handle the case where no restart happens.

File: crates/core/tedge_api/src/messages.rs
Comment (Bravo555) [inline crates/core/tedge_api/src/messages.rs:603]: Is the `reason` field supposed to be not empty only when `status` is `CommandStatus::Failed`? If there are no other motivations for this, I suggest using [Serde's internally tagged enum representation](https://serde.rs/enum-representations.html#internally-tagged). Modified `CommandStatus` would look like this:

```rust
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Clone)]
#[serde(rename_all = "camelCase", tag = "status")]
pub enum CommandStatus {
    Init,
    Executing,
    Successful,
    Failed { reason: String },
}
```

And the `reason` field would automatically be serialised/deserialised and only be present when in `Failed` variant, which looks the same as the current structure.

One issue is that `CommandStatus` would no longer be copy, and also perhaps there would have to be more code changes.

`RestartCommandPayload would then simply be:

```suggestion
#[derive(Debug, Clone, Deserialize, Serialize, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RestartCommandPayload {
    #[serde(flatten)]
    pub status: CommandStatus,
}
```

**Review by reubenmiller — CHANGES_REQUESTED** [review]
The PRs looking very close now, I pushed some minor fixes and added some more system tests.

Though only remaining minor change is the following:

* Log entry is misleading when the restart command is killed by a signal. The entry with `Restart command has been interrupted by a signal` is good, but it is proceeded by an entry with the opposite meaning `The restart command has been successfully executed`.

    ```
    Sep 21 21:30:08 d2dcf7b3095f tedge-agent[189]: 2023-09-21T21:30:08.072623262Z  INFO tedge_agent::restart_manager::actor: Restart command has been interrupted by a signal: Command { std: "/usr/bin/sudo" "/usr/bin/on_shutdown_signal_interrupt.sh", kill_on_drop: false }
    Sep 21 21:30:08 d2dcf7b3095f tedge-agent[189]: 2023-09-21T21:30:08.072724928Z  INFO tedge_agent::restart_manager::actor: The restart command has been successfully executed
    ```

**Review by albinsuresh — COMMENTED** [review]
The code looks fine except for a small issue that I've highlighted around failed state repository updates, that's not handled cleanly(even in the old code). Happy to approve once that's fixed.

File: crates/core/tedge_agent/src/restart_manager/actor.rs
Comment (albinsuresh) [reply crates/core/tedge_agent/src/restart_manager/actor.rs:90]: > From his perspective, the successful execution of the script should have translated to a Success, right?

I guess this is the root of the problem. When it's a user-overridden script, we can't always assume that the device would have been restarted, and wait for the agent to reboot and really validate if the restart succeeded and then only send the `Successful` status. But, purely relying on the exit status is not sufficient for our default restart script either. So, it's like a chicken and egg problem.

If we can always assume that the user-provided scripts must restart the device, then this is fine. Else we should expect their custom script to send the `successful` status update MQTT message as well, at the end of their "non-restarting" logic.

File: crates/core/tedge_agent/src/tedge_operation_converter/actor.rs
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/tedge_operation_converter/actor.rs:118]: No this is temporary code. The first part of the method is related to `tedge` topics, the second part to `te` topics.

I could have write this differently, but came with this not so nice return to avoid confusing diffs in the PR.

**Comment by didier-wenzek** [comment]
>    * Log entry is misleading when the restart command is killed by a signal.
The entry with `Restart command has been interrupted by a signal` is good,
but it is proceeded by an entry with the opposite meaning `The restart command has been successfully executed`.

Addressed by https://github.com/thin-edge/thin-edge.io/pull/2245/commits/6cb2de650fcc4446512a93bece631817b829c21e

File: crates/core/tedge_agent/src/restart_manager/actor.rs
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/restart_manager/actor.rs:58]: Fixed by https://github.com/thin-edge/thin-edge.io/pull/2245/commits/97a07a0e33a30e6ef745f71053f3cfecc67d8e96

File: crates/core/tedge_agent/src/restart_manager/actor.rs
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/restart_manager/actor.rs:99]: > What if the user intentionally replaced the restart script with something that doesn't really restart the device but does something else?

The user should not be surprised that the *restart*  command fails ;-)

> If we can always assume that the user-provided scripts must restart the device, then this is fine.

We can and we do. Even more, we check that this is the case: if the agent doesn't shutdown than restart after a reboot of the device, then command execution is reported as failed.

File: tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot
Comment (didier-wenzek) [reply tests/RobotFramework/tests/cumulocity/restart/restart_device_child.robot]: fixed

File: crates/core/tedge_api/src/messages.rs
Comment (didier-wenzek) [reply crates/core/tedge_api/src/messages.rs]: fixed

File: crates/core/tedge_api/src/messages.rs
Comment (didier-wenzek) [reply crates/core/tedge_api/src/messages.rs]: fixed

**Review by reubenmiller — APPROVED** [review]
The requested log message changes were resolved.

Approved

**Pull request merged**
